<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

In [2]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgres://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [5]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select count(store) from stores

"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**: 1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [14]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select name from stores
order by name asc
"""

pd.read_sql_query(sql, engine)

name
0     218 Fuel Express & Chubby's Liquor
1             3047 Cub Foods / Iowa City
2                  3060 Cub Foods / Ames
3            3061 Cub Foods / Sioux City
4                     3rd St Convenience
...                                  ...
1968       Xo Food And Liquor / Waterloo
1969                       Xpress Liquor
1970                      Yr Dollar Stop
1971                Zapf's Pronto Market
1972                      Z's Quickbreak

[1973 rows x 1 columns]

**Prompt 2 Answer**: '3061 Cub Foods / Sioux City'

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [50]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select * from sales
where category_name = 'IMPORTED VODKA'
order by vendor
limit 3
"""

pd.read_sql_query(sql, engine)

date convenience_store  store county_number     county category  \
0 2014-07-01              None   4509            85      Story  1032080   
1 2014-10-08              None   4046             4  Appanoose  1032080   
2 2014-06-19              None   4481            52    Johnson  1032080   

    category_name vendor_no                     vendor    item  \
0  IMPORTED VODKA        14  Adamba Imports Int'l Inc.  904783   
1  IMPORTED VODKA        10      A Hardy / U.S.A. Ltd.   36200   
2  IMPORTED VODKA        10      A Hardy / U.S.A. Ltd.   36200   

                   description  pack  liter_size state_btl_cost btl_price  \
0                 Vesica Vodka    12         750          $5.99     $8.99   
1  Michael Godard "Mr. Smooth"     6         750         $19.00    $28.50   
2  Michael Godard "Mr. Smooth"     6         750         $19.00    $28.50   

   bottle_qty   total  
0          12  107.88  
1           1   28.50  
2           3   85.50

**Prompt 3 Answer**: 

1: price 8.99, qty 12

2: price 28.50, qty 1

3: price 28.50, qty 1

### Prompt 4

What is the total amount of liquor sold in February 2015?

In [75]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select sum(pack) from sales 
where date > '2015-02-01' and date <'2015-02-28'
limit 5


"""
pd.read_sql_query(sql, engine)

sum
0  1937899

**Prompt 4 Answer**: 

1,591,962 bottles

1,937,899 packs

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [108]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select vendor as vendor, count(*) as transaction_count, sum(bottle_qty) as num_bottles ,sum(total) as total
from sales
where category_name = 'IMPORTED VODKA'
group by vendor
order by transaction_count DESC
limit 4
"""

pd.read_sql_query(sql, engine)

vendor  transaction_count  num_bottles  \
0  Pernod Ricard USA/Austin Nichols              42338       403450   
1               Bacardi U.S.A. Inc.              26441       236047   
2                   Diageo Americas              15725       152038   
3   Constellation Wine Company Inc.              12565       116053   

        total  
0  7880365.88  
1  5771242.07  
2  3346425.42  
3  1895005.51

**Prompt 5 Answer**: 

*Pernod Ricard USA/Austin Nichols*

42,338 transactions

403,450 bottles

78,80,365.88 revenue

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [139]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select btl_price - state_btl_cost AS markup
FROM sales

limit 5
"""

pd.read_sql_query(sql, engine)

markup
0  $5.75
1  $5.75
2  $5.75
3  $8.58
4  $2.50

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [141]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select store, avg((btl_price-state_btl_cost)/state_btl_cost) as percentage_markup
from sales
group by store
order by percentage_markup DESC
limit 5
"""

pd.read_sql_query(sql, engine)

store  percentage_markup
0   4013           0.525471
1   4024           0.521499
2   4266           0.518903
3   3833           0.515094
4   3660           0.514591

**Prompt 7 Answer**: *store 4013 has the highest average markup of about 52%*

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [159]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT st.store AS store_ID, 
min((btl_price-state_btl_cost)/state_btl_cost) AS min_markup,
max((btl_price-state_btl_cost)/state_btl_cost) AS max_markup,
avg((btl_price-state_btl_cost)/state_btl_cost) AS avg_percentage_markup, 
st.name AS name, 
st.store_address
FROM sales ss
JOIN stores st on st.store = ss.store
GROUP BY st.store
ORDER BY avg_percentage_markup DESC
LIMIT 5

"""

pd.read_sql_query(sql, engine)

store_id  min_markup  max_markup  avg_percentage_markup  \
0      4013    0.498571    6.708763               0.525471   
1      4024    0.498462    6.708763               0.521499   
2      4266    0.498462    6.708763               0.518903   
3      3833    0.498462    6.708763               0.515094   
4      3660    0.498462    6.708763               0.514591   

                         name  \
0      Wal-Mart 0841 / Tipton   
1  Wal-Mart 1546 / Iowa Falls   
2  Wal-Mart 1683 / Shenandoah   
3    Wal-Mart 3394 / Atlantic   
4   Wal-Mart 2935 / Knoxville   

                                       store_address  
0  1126 Highway 38 North\nTipton, IA 527720000\n(...  
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...  
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...  
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....  
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...

**Prompt 8 Answer**:

Wal-Mart 0841 / Tipton

Wal-Mart 1546 / Iowa Falls

Wal-Mart 1683 / Shenandoah

Wal-Mart 3394 / Atlantic

Wal-Mart 2935 / Knoxville

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

<img src="./images/schema.png" alt="schema" width="750"/>

In [187]:
# Use a SQL query to find the answer to the above prompt.

sql = """

select c.county, sum(bottle_qty)/c.population as bottles_per_person, round(avg(liter_size)) as avg_bottle_size from counties c
join sales s on c.county = s.county
group by c.county
order by bottles_per_person DESC, avg_bottle_size DESC
limit 9

"""

pd.read_sql_query(sql, engine)

county  bottles_per_person  avg_bottle_size
0      Dickinson                  22            974.0
1           Polk                  15            892.0
2     Black Hawk                  15            861.0
3    Cerro Gordo                  14            986.0
4        Johnson                  13            895.0
5          Scott                  13            870.0
6           Linn                  12            885.0
7        Kossuth                  11           1041.0
8  Pottawattamie                  11            891.0

**Prompt 9 Answer**:

Kossuth County

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [216]:
# Use a SQL query to find the answer to the above prompt.

sql = """
select cast(total as money)-(btl_price*bottle_qty) as test_total
from sales
order by test_total DESC
limit 5

"""

pd.read_sql_query(sql, engine)

test_total
0      $0.00
1      $0.00
2      $0.00
3      $0.00
4      $0.00

**Prompt 10 Answer**: